src/Microsoft.OData.Client/BaseSaveResult.cs

```diff
@@ -1104,8 +1104,26 @@ private static Uri AppendTargetEntityKeyIfNeeded(Uri linkUri, LinkDescriptor bin
 
             Debug.Assert(targetResource != null, "targetResource != null");
             StringBuilder builder = new StringBuilder();
-            builder.Append(UriUtil.UriToString(linkUri));
-            builder.Append(UriHelper.QUESTIONMARK + XmlConstants.HttpQueryStringId + UriHelper.EQUALSSIGN + targetResource.Identity);
+            string uriString = UriUtil.UriToString(linkUri);
+
+            if (this.RequestInfo.Context.DeleteLinkUriOption == DeleteLinkUriOption.RelatedKeyAsSegment)
+            {
+                // Related key segment should appear before /$ref
+                int indexOfLinkSegment = uriString.IndexOf(XmlConstants.UriLinkSegment, StringComparison.Ordinal);
+                builder.Append(indexOfLinkSegment > 0 ? uriString.Substring(0, indexOfLinkSegment - 1) : uriString);

```

Comment:
why not `>= 0` since 0 can be a valid index?

Reply:
Maybe we can use `Span` and slice the string to avoid allocation with `Substring`

Reply:
Using `>= 0` would suggest that `uriString` starts with `$ref` which is not practical in this case. So while we can make that change, it's really not a practical scenario

Reply:
@habbes `Span` (and `ReadOnlySpan`) are not available in NET45 but we could introduce the `System.Memory` package to take care of that. Maybe to keep the change small we could live with that single extra allocation since this code is not on a hot path